sql-plugin/src/main/scala/org/apache/spark/sql/rapids/mathExpressions.scala

```diff
@@ -453,13 +453,160 @@ abstract class GpuRoundBase(child: Expression, scale: Expression) extends GpuBin
+    // type. Under this circumstance, cuDF may produce different results to Spark.
+    //
+    // In this method, we handle round overflow, aligning the inconsistent results to Spark.
+    //
+    // For scales exceeding max digits, we can simply return zero values.
+    //
+    // For scales equaling to max digits, we need to perform round. Fortunately, round up
+    // will NOT occur on the max digits of numeric types except LongType. Therefore, we only
+    // need to handle round down for most of types, through returning zero values.
+    def fixUpOverflowInts(zeroFn: () => Scalar): ColumnVector = {
+      val scaleVal = scale.getValue.asInstanceOf[Int]
+
+      // Rounding on the max digit of long values, which should be specialized handled since
+      // it may be needed to round up, which will produce inconsistent results because of
+      // overflow. Otherwise, we only need to handle round down situations.
+      if (-scaleVal == 19 && lhsValue.getType == DType.INT64) {
+        fixUpInt64OnBounds(zeroFn)
+      } else if (-scaleVal >= DecimalUtil.getPrecisionForIntegralType(lhsValue.getType)) {
+        withResource(zeroFn()) { s =>
+          withResource(ColumnVector.fromScalar(s, lhsValue.getRowCount.toInt)) { zero =>
+            // set null mask if necessary
+            if (lhsValue.hasNulls) {
+              zero.mergeAndSetValidity(BinaryOp.BITWISE_AND, lhsValue)
+            } else {
+              zero.incRefCount()
+            }
+          }
+        }
+      } else {
+        lhsValue.round(scaleVal, roundMode)
+      }
+    }
+
+    // Compared to other non-decimal numeric types, Int64(LongType) is a bit special in terms of
+    // rounding by the max digit. Because the bound values of LongType can be rounded up, while
+    // other numeric types can only be rounded down:
+    //
+    //  the max value of Byte: 127
+    //  The first digit is up to 1, which can't be rounded up.
+    //  the max value of Short: 32767
+    //  The first digit is up to 3, which can't be rounded up.
+    //  the max value of Int32: 2147483647
+    //  The first digit is up to 2, which can't be rounded up.
+    //  the max value of Float32: 3.4028235E38
+    //  The first digit is up to 3, which can't be rounded up.
+    //  the max value of Float64: 1.7976931348623157E308
+    //  The first digit is up to 1, which can't be rounded up.
+    //  the max value of Int64: 9223372036854775807
+    //  The first digit is up to 9, which can be rounded up.
+    //
+    // When rounding up 19-digits long values on the first digit, the result can be 1e19 or -1e19.
+    // Since LongType can not hold these two values, the 1e19 overflows as -8446744073709551616L,
+    // and the -1e19 overflows as 8446744073709551616L. The overflow happens in the same way for
+    // HALF_UP (round) and HALF_EVEN (bround).
+    def fixUpInt64OnBounds(zeroFn: () => Scalar): ColumnVector = {
```

Comment:
Here only handles the long type, so seems this `zeroFn` is not necessary.

It should always be a zero column of the long type, right ?

BTW, do we really need to take care of this round up/down of long ? Can cudf `round` handle it the same as what you are doing here ?

Reply:
You are right. I removed the argument `zeroFn`. 
When rounding by scales exceeding the max digits, the results of `cudf::round` are inconsistent with Spark. And in terms of cuDF, it is a undefined behavior. So, technically cuDF can return any value under this situation. That's why we need to override the round of extreme large scales.